### PR TITLE
Add comment about Thanos, external labels and downsampling.

### LINKED ISSUE #2097: Document how metaconvert treats Thanos external labels
#### Describe the bug

Documentation on `metaconvert` (https://grafana.com/docs/mimir/latest/migration-guide/migrating-from-thanos-or-prometheus/) doesn't mention what happens to external labels used by Thanos

We should also mention how to handle blocks with duplicated data coming from Thanos. Thanos can deduplicate using replica label (`--query.replica-label`), but uploading blocks from all replicas into Mimir and running `metaconvert` would create problems, as `metaconvert` removes labels, and Mimir compactor would then merge all such blocks together.

Pointers:

- https://thanos.io/v0.2/thanos/getting-started.md/#query-layercomponentsquerymd
- https://thanos.io/v0.2/thanos/getting-started.md/#deduplicating-data-from-prometheus-ha-pairs

### PR DESCRIPTION
**Pull request by pstibrany**
#### What this PR does

This PR documents how external labels stored in `meta.json` files of blocks coming from Thanos interact with Mimir.

After writing this, I think we should deprecate `metaconvert` tool completely, as it no longer servers its original purpose – add `__org_id__` label to the blocks. Now it only removes external labels, which can be actively harmful: compactor will merge blocks with no labels, and possibly generate blocks with duplicated samples.

If user preserved the labels instead, at least the blocks would not get merged together (because compactor still respects them), although Mimir still wouldn't return all the labels on the output.

**Update:** mentions of `metaconvert` were removed from the documentation. Instead we now explain how Mimir uses external labels that are present in blocks generated by Thanos.

#### Which issue(s) this PR fixes or relates to

Fixes #2097 

#### Checklist

- [na] Tests updated
- [na] Documentation added
- [na] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by pracucci** [comment]
> After writing this, I think we should deprecate metaconvert tool completely, as it no longer servers its original purpose – add __org_id__ label to the blocks. Now it only removes external labels, which can be actively harmful: compactor will merge blocks with no labels, and possibly generate blocks with duplicated samples.

Good point. What if we go on and just update the migration doc, without mentioning `metaconvert` at all? We could just add that the migration doc requires Mimir 2.1.0 or above.

**Comment by pstibrany** [comment]
> Good point. What if we go on and just update the migration doc, without mentioning `metaconvert` at all? We could just add that the migration doc requires Mimir 2.1.0 or above.

Sounds good to me.

File: docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md [hunk 1]
```diff
@@ -47,53 +47,33 @@ aws s3 cp <DIRECTORY> s3://<TENANT>/<DIRECTORY>
...
-   ```bash
-   docker run grafana/metaconvert -backend=gcs -gcs.bucket-name=bucket -tenant=anonymous -dry-run
-   ```
+Thanos requires that Prometheus is configured with external labels.
+When Thanos sidecar uploads blocks, it includes these external labels into `meta.json` file inside the block.
+Thanos then returns these external labels when block is queried, and can also use labels for deduplication of replicated data.
 
-   - If you are using a local binary, run the following command:
+Mimir doesn't include external labels in the query results.
+This means that blocks that were originally created by Thanos will miss all of their external labels in the results when queried.
```
Comment (56quarters) [inline docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md:72]:
```suggestion
This means that blocks that were originally created by Thanos will not include their external labels in the results when queried.
```

File: docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md [hunk 2]
```diff
@@ -47,53 +47,33 @@ aws s3 cp <DIRECTORY> s3://<TENANT>/<DIRECTORY>
...
-
-> **Warning:** The `metaconvert` tool modifies objects in place.
-> Ensure you enable bucket versioning or have backups before running the tool.
-
-To run `metaconvert`, provide it with the bucket configuration. Use `metaconvert -h` to list available parameters.
+When querying blocks, Mimir ignores all external labels and doesn't add them to the results.
+During compaction Mimir will respect external labels in the block and only merge blocks with the same labels together.
 
-1. Complete one of the following steps to run `metaconvert` in a dry-run mode, which lists blocks for migration.
+Note that blocks from Prometheus do not have any labels stored in them, only blocks from Thanos use them.
```
Comment (56quarters) [inline docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md:63]: (not sure if my suggestion is correct, but if this is what you meant I think it's less ambiguous)
```suggestion
Note that blocks from Prometheus do not have any external labels stored in them, only blocks from Thanos use them.
```

File: docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md [hunk 3]
```diff
@@ -47,53 +47,33 @@ aws s3 cp <DIRECTORY> s3://<TENANT>/<DIRECTORY>
...
-   ```bash
-   ./metaconvert -backend=filesystem -filesystem.dir=/bucket -tenant=anonymous -dry-run
-   ```
+Thanos can also be configured to use labels for deduplication. After uploading such blocks into Mimir, query results will contain duplicate samples (assuming they don't use exact same timestamp).
+Mimir's Compactor will however not compact such blocks together, as long as blocks have different set of external labels.
 
-1. Remove the `-dry-run` flag to apply the migration.
-1. Verify the migration by re-running the tool with `-dry-run` and confirming that no changes are required.
+Mimir also doesn't understand Thanos "downsampling" feature, that is `Resolution` field from `meta.json` file created by Thanos is ignored by Mimir.
+Notable exception is a compactor, which will not compact blocks with different resolutions together.
```
Comment (56quarters) [inline docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md:78]:
```suggestion
A notable exception is the Mimir compactor, which will not compact blocks with different resolutions together.
```

**Review by 56quarters — APPROVED** [review]

**Review by pracucci — COMMENTED** [review]
I realized I didn't clicked the "submit review" button when I did review it few days ago 🤦

File: docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md [hunk 4]
```diff
@@ -47,53 +47,33 @@ aws s3 cp <DIRECTORY> s3://<TENANT>/<DIRECTORY>
...
 
-## Migrating the block `meta.json` metadata using `metaconvert`
+## Block metadata
 
 Every block has a `meta.json` metadata file used by Grafana Mimir, Prometheus, and Thanos to identify the block contents.
 Each project has its own metadata conventions.
-The `metaconvert` tool migrates the `meta.json` metadata from one project to another.
 
-### Downloading `metaconvert`
+Ingesters from Mimir 2.1 and earlier added an external label into `meta.json` file to identify the tenant that owns the block.
```
Comment (osg-grafana) [inline docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md:55]:
```suggestion
From the Grafana Mimir 2.1 release or earlier, the ingesters added an external label into the `meta.json` file to identify the tenant that owns the block.
```

File: docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md [hunk 5]
```diff
@@ -47,53 +47,33 @@ aws s3 cp <DIRECTORY> s3://<TENANT>/<DIRECTORY>
...
+Mimir compactor also adds special label identifying compactor-shard when split-and-merge sharding feature is enabled.
```
[the remark this answers is not in the document]
Comment (osg-grafana) [reply docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md:56]: @pstibrany, I assume a human is performing this action. Passive voice makes that ambiguous. Please vet for accuracy.

File: docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md [hunk 6]
```diff
@@ -47,53 +47,33 @@ aws s3 cp <DIRECTORY> s3://<TENANT>/<DIRECTORY>
...
 
-- If you are using Docker, run the following command:
+In Mimir 2.2 and later blocks no longer have a label identifying the tenant. Compactor shard label is still used.
```
[the remark this answers is not in the document]
Comment (osg-grafana) [reply docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md:58]: This information sounds like information that ought to be in the release notes rather than the core documentation.

File: docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md (hunk 4 above)
Comment (osg-grafana) [reply docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md:55]: This information sounds like information that ought to be in the release notes rather than the core documentation.

File: docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md [hunk 7]
```diff
@@ -47,53 +47,33 @@ aws s3 cp <DIRECTORY> s3://<TENANT>/<DIRECTORY>
...
-go install github.com/grafana/mimir/cmd/metaconvert@latest
-```
-
-### Running `metaconvert`
-
-> **Warning:** The `metaconvert` tool modifies objects in place.
-> Ensure you enable bucket versioning or have backups before running the tool.
-
-To run `metaconvert`, provide it with the bucket configuration. Use `metaconvert -h` to list available parameters.
+When querying blocks, Mimir ignores all external labels and doesn't add them to the results.
```
Comment (osg-grafana) [inline docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md:60]:
```suggestion
When querying blocks, Mimir ignores all external labels and doesn't add them to the results.
```
What does "When querying blocks" mean?
* Does it get blocked (by what?)
* Does it block something? In either case, who or what, and by whom?

File: docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md [hunk 8]
```diff
@@ -47,53 +47,33 @@ aws s3 cp <DIRECTORY> s3://<TENANT>/<DIRECTORY>
...
+During compaction Mimir will respect external labels in the block and only merge blocks with the same labels together.
```
Comment (osg-grafana) [inline docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md:61]:
```suggestion
During compaction, Mimir respects the external labels in the block and only merges blocks with the same labels.
```

File: docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md [hunk 9]
```diff
@@ -47,53 +47,33 @@ aws s3 cp <DIRECTORY> s3://<TENANT>/<DIRECTORY>
...
 
-1. Complete one of the following steps to run `metaconvert` in a dry-run mode, which lists blocks for migration.
+Note that blocks from Prometheus do not have any external labels stored in them, only blocks from Thanos use them.
```
Comment (osg-grafana) [inline docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md:63]:
```suggestion
> **Note:** Blocks from Prometheus do not have any external labels stored in them; only blocks from Thanos use labels.
```

File: docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md [hunk 10]
```diff
@@ -47,53 +47,33 @@ aws s3 cp <DIRECTORY> s3://<TENANT>/<DIRECTORY>
...
 
-   - If you are using Docker, run the following command:
+### Note about Thanos: external labels and downsampling
```
Comment (osg-grafana) [inline docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md:65]:
```suggestion
```

File: docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md [hunk 11]
```diff
@@ -47,53 +47,33 @@ aws s3 cp <DIRECTORY> s3://<TENANT>/<DIRECTORY>
...
 
-   ```bash
-   docker run grafana/metaconvert -backend=gcs -gcs.bucket-name=bucket -tenant=anonymous -dry-run
-   ```
+Thanos requires that Prometheus is configured with external labels.
+When Thanos sidecar uploads blocks, it includes these external labels into `meta.json` file inside the block.
+Thanos then returns these external labels when block is queried, and can also use labels for deduplication of replicated data.
```
Comment (osg-grafana) [inline docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md:69]:
```suggestion
> **Note:** Thanos requires that Prometheus is configured with external labels.
When the Thanos sidecar uploads blocks, it includes the external labels from Prometheus in the `meta.json` file inside the block.
When you query the block, Thanos returns Prometheus’ external labels. Thanos also uses labels for the deduplication of replicated data.
```
As a fellow human, am I querying the block or is something doing that for me?

File: docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md [hunk 12]
```diff
@@ -47,53 +47,33 @@ aws s3 cp <DIRECTORY> s3://<TENANT>/<DIRECTORY>
...
 
-   - If you are using a local binary, run the following command:
+Mimir doesn't include external labels in the query results.
+This means that blocks that were originally created by Thanos will not include their external labels in the results when queried.
```
Comment (osg-grafana) [inline docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md:72]:
```suggestion
Blocks that were originally created by Thanos do not include their external labels in the query results, because Mimir doesn't include external labels in the query results.
```

File: docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md [hunk 13]
```diff
@@ -47,53 +47,33 @@ aws s3 cp <DIRECTORY> s3://<TENANT>/<DIRECTORY>
...
 
-   ```bash
-   ./metaconvert -backend=filesystem -filesystem.dir=/bucket -tenant=anonymous -dry-run
-   ```
+Thanos can also be configured to use labels for deduplication. After uploading such blocks into Mimir, query results will contain duplicate samples (assuming they don't use exact same timestamp).
+Mimir's Compactor will however not compact such blocks together, as long as blocks have different set of external labels.
```
Comment (osg-grafana) [inline docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md:75]:
```suggestion
You can configure Thanos to use labels for deduplication. After uploading such blocks into Mimir, the query results will contain duplicate samples (assuming their timestamps are not identical).
As long as the blocks have different set of external labels, Mimir's compactor does not compact such blocks together.
```

File: docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md [hunk 14]
```diff
@@ -47,53 +47,33 @@ aws s3 cp <DIRECTORY> s3://<TENANT>/<DIRECTORY>
...
 
-1. Remove the `-dry-run` flag to apply the migration.
-1. Verify the migration by re-running the tool with `-dry-run` and confirming that no changes are required.
+Mimir also doesn't understand Thanos "downsampling" feature, that is `Resolution` field from `meta.json` file created by Thanos is ignored by Mimir.
+A notable exception is the Mimir compactor, which will not compact blocks with different resolutions together.
+Blocks generated by Mimir don't set the resolution field at all.
```
Comment (osg-grafana) [inline docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md:79]:
```suggestion
> **Note:** Mimir does not understand Thanos’ _downsampling_ feature. As a result, Mimir ignore the `Resolution` field from the `meta.json` file that Thanos creates.
A notable exception is that the Mimir compactor does not compact together blocks of different resolutions.
The blocks that Mimir generates have unset resolution fields.
```

File: docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md [hunk 15]
```diff
@@ -51,7 +51,34 @@ gsutil -m cp -r <DIRECTORY> gs://<TENANT>/<DIRECTORY>
...
+Mimir doesn't support external labels feature, and only uses few labels internally. `metaconvert` will remove all external labels from `meta.json` file, and only keep labels used by Mimir, if they are valid.
+
+This means that blocks that were originally created by Thanos will miss some of the labels when queried.
+
+If Thanos system was using labels for deduplication, and all such blocks are uploaded to Mimir's bucket, after running `metaconvert` and removing the labels, Mimir will compact such blocks together.
+This will result in blocks with duplicated samples (unless sample timestamps were exactly the same).
+
+Mimir also doesn't understand Thanos "downsampling" feature.
+While `metaconvert` currently doesn't remove `Resolution` field from `meta.json` file, the rest of Mimir does not respect the field during query time.
+Notable exception is a compactor, which will not compact blocks with different resolutions together.
```
[the remark this answers is not in the document]
Comment (pstibrany) [reply docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md]: Based on ealier feedback, this now reads:

```
A notable exception is the Mimir compactor, which will not compact blocks with different resolutions together.
```

WDYT?

File: docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md (hunk 4 above)
Comment (pstibrany) [reply docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md:55]: It is mentioned in the Mimir 2.2 release notes as:

```
* [CHANGE] Blocks uploaded by ingester no longer contain `__org_id__` label. Compactor now ignores this label and will compact blocks with and without this label together. `mimirconvert` tool will remove the label from blocks as "unknown" label. #1972
```

File: docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (pstibrany) [reply docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md:56]: Thanks for suggestion. I would rephrase it slightly to associate "sharding feature" with "split-and-merge compactor".

```suggestion
If you enable the sharding feature of Mimir's split-and-merge compactor, compactor adds a special label to identify the compactor-shard.
```

WDYT?

File: docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (pstibrany) [reply docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md:58]: > What technical value does "still" have here?

It is the only label left that Mimir (still) uses internally. All other internal labels are not used by Mimir anymore.

File: docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md (hunk 7 above)
Comment (pstibrany) [reply docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md:60]: > What does "When querying blocks" mean?
> 
> * Does it get blocked (by what?)
> * Does it block something? In either case, who or what, and by whom?

"blocks" is used as a noun here, not verb. When user sends a query to execute, we say that "Mimir queries (TSDB) blocks".

File: docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md (hunk 12 above)
Comment (pstibrany) [reply docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md:72]: Without "Mimir doesn't include external labels in the query results.", the sentence doesn't mention that this happens when Mimir uses these blocks from Thanos.

Comment (pstibrany) [reply docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md:72]: But since previous paragraph is a "note about thanos", then it shows that this one is about Mimir. I think that works.

**Comment by pstibrany** [comment]
Thanks @pracucci and @osg-grafana for your reviews. I have applied your suggestions. Please take a look again when you have time.

File: docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md [hunk 16]
```diff
@@ -51,7 +51,34 @@ gsutil -m cp -r <DIRECTORY> gs://<TENANT>/<DIRECTORY>
...
+
+In Mimir 2.1 and earlier, `metaconvert` added external label into `meta.json` file to identify the tenant that owns the block.
+It also removed all other labels, only preserving labels known to Mimir.
+
+In Mimir 2.2 and later blocks no longer have a label identifying the tenant, and `metaconvert` simply removes all external labels from `meta.json` file not known to Mimir.
+
+Mimir itself does not use labels from `meta.json` file. Note that blocks from Prometheus do not have any labels stored in them,
+only blocks from Thanos use them.
+
+### Note about Thanos: external labels and downsampling
```
[the remark this answers is not in the document]
Comment (pstibrany) [reply docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md]: Now it's:

```
> **Note:** Mimir does not support Thanos’ _downsampling_ feature. As a result, Mimir ignores the `Resolution` field from the `meta.json` file that Thanos creates.
```

File: docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md [hunk 17]
```diff
@@ -47,53 +47,32 @@ aws s3 cp <DIRECTORY> s3://<TENANT>/<DIRECTORY>
 gsutil -m cp -r <DIRECTORY> gs://<TENANT>/<DIRECTORY>
 ```
 
-## Migrating the block `meta.json` metadata using `metaconvert`
+## Block metadata
```
[the remark this answers is not in the document]
Comment (pstibrany) [reply docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md:50]: I get what you're saying. I think the answer right now depends on your needs:
- if you don't mind losing external labels in results, you can upload blocks as-is, and forget about them
- if you used deduplication, you most likely only want to upload blocks from one replica, but not the other
- if you DO mind losing external labels, currently we don't have a solution. We will one partial solution in backfilling API – writing external labels into series, if they don't conflict with existing labels.

Shall we be explicit about this?

[the remark this answers is not in the document]
Comment (pracucci) [reply docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md:50]: Yes. I think that's exactly what you should document in a user-facing doc, not the reason behind it, the difference between versions and so on. Keep in mind doc is versioned, so I think we should just address 2.2.0.

[the remark this answers is not in the document]
Comment (pstibrany) [reply docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md:50]: I have introduced `## Considerations on Thanos specific features` section and completely rephrased it to be explicit about what we recommend in different scenarios. Please take a look again.

**Comment by pstibrany** [comment]
@pracucci @osg-grafana I believe I've addressed your feedback, can you please take another look?

File: docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md [hunk 18]
```diff
@@ -47,53 +47,35 @@ aws s3 cp <DIRECTORY> s3://<TENANT>/<DIRECTORY>
...
 
-Every block has a `meta.json` metadata file used by Grafana Mimir, Prometheus, and Thanos to identify the block contents.
+Each block has a `meta.json` metadata file that is used by Grafana Mimir, Prometheus, and Thanos to identify the block contents.
 Each project has its own metadata conventions.
-The `metaconvert` tool migrates the `meta.json` metadata from one project to another.
 
-### Downloading `metaconvert`
+In the Grafana Mimir 2.1 (or earlier) release, the ingesters added an external label to the `meta.json` file to identify the tenant that owns the block.
+If you enable the sharding feature of Mimir's split-and-merge compactor, compactor adds a special label to identify the compactor-shard.
```
Comment (pracucci) [inline docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md:56]: This sentence looks like related to 2.1 or earlier only.

File: docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md [hunk 19]
```diff
@@ -47,53 +47,35 @@ aws s3 cp <DIRECTORY> s3://<TENANT>/<DIRECTORY>
...
-> Ensure you enable bucket versioning or have backups before running the tool.
-
-To run `metaconvert`, provide it with the bucket configuration. Use `metaconvert -h` to list available parameters.
+> **Note**: Blocks from Prometheus do not have any external labels stored in them; only blocks from Thanos use labels.
 
-1. Complete one of the following steps to run `metaconvert` in a dry-run mode, which lists blocks for migration.
+## Considerations on Thanos specific features
 
-   - If you are using Docker, run the following command:
+> **Note**: Thanos requires that Prometheus is configured with external labels.
```
Comment (pracucci) [inline docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md:64]: Looks a bit weird to me starting the section with a "note" (so far I've seen notes being addendum at the end of a section). I would consider making these sentences normal ones, not a "note".

File: docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md [hunk 20]
```diff
@@ -47,53 +47,35 @@ aws s3 cp <DIRECTORY> s3://<TENANT>/<DIRECTORY>
...
-   - If you are using a local binary, run the following command:
+**Grafana Mimir doesn't inject external labels into query results.**
+This means that blocks that were originally created by Thanos will not include their external labels in the results when queried.
+If you need to have external labels in your query results, this is currently not possible to achieve in Grafana Mimir.
 
-   ```bash
-   ./metaconvert -backend=filesystem -filesystem.dir=/bucket -tenant=anonymous -dry-run
-   ```
+**Grafana Mimir will not respect deduplication labels configured in Thanos when querying the blocks.**
+For the best results please only upload Thanos blocks from single Prometheus replica from each HA pair.
```
Comment (pracucci) [inline docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md:75]: What does "best results" mean here? If I upload all Thanos blocks, are query results incorrect or just less performant? I think they're just less performant, so I would clarify it ("For the best query performance...").

File: docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md [hunk 21]
```diff
@@ -47,53 +47,35 @@ aws s3 cp <DIRECTORY> s3://<TENANT>/<DIRECTORY>
...
+If you upload blocks from both replicas, the query results returned by Mimir will include samples from both replicas.
 
-1. Remove the `-dry-run` flag to apply the migration.
-1. Verify the migration by re-running the tool with `-dry-run` and confirming that no changes are required.
+**Grafana Mimir does not support Thanos’ _downsampling_ feature.**
+For best results please only upload original (raw) Thanos blocks into Mimir's storage.
```
Comment (pracucci) [inline docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md:79]: In this case "best results" means "To guarantee query results correctness". I would clarify it.

**Review by pracucci — APPROVED** [review]
Thanks for addressing my feedback! It's way better now. I left final comments.

File: docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md (hunk 18 above)
Comment (pstibrany) [reply docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md:56]: Next paragraph says it still applies for 2.2:

> The compactor-shard label is still used, if sharding of blocks during compaction is enabled.

File: docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md (hunk 20 above)
Comment (pstibrany) [reply docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md:75]: Good point.

File: docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md (hunk 21 above)
Comment (pstibrany) [reply docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md]: Clarified.

File: docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md (hunk 18 above)
Comment (pracucci) [reply docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md:56]: I know but people sometimes just stop reading. Given the split-and-merge compactor shard label doesn't change by Mimir versions, I wouldn't mention it along with 2.1 or 2.2 changes at all.

Comment (pstibrany) [reply docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md:56]: > Given the split-and-merge compactor shard label doesn't change by Mimir versions, I wouldn't mention it along with 2.1 or 2.2 changes at all.

OK.

File: docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (FUSAKLA) [reply docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md:50]: Hi, just passing by, thinking about how to do migration from Thanos. 

Few points:
 - If you want to preserve external labels, you should in theory be able to use the `thanos bucket rewrite --rewrite.to-relabel-config="..."` to rewrite all the blocks and add te external labels to it. Thah could work I hope? See https://thanos.io/tip/components/tools.md/#bucket-rewrite
 - And about deduplication deduplication, there is possibility to use vertical compaction with deduplication first, using something like  `thanos compact --compact.enable-vertical-compaction --deduplication.func=penalty --deduplication.replica-label=LABEL` see https://thanos.io/tip/components/compact.md/#vertical-compaction-use-cases 
 
Might be worth adding note about those?

[the remark this answers is not in the document]
Comment (pstibrany) [reply docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md:50]: Thank you @FUSAKLA! That's very useful, I will add it to the PR.

**Comment by pstibrany** [comment]
@osg-grafana any chance for re-review anytime soon? I would like to finish this in time for 2.2 release.

**Review by osg-grafana — APPROVED** [review]
Unblocking and will look at this again when I have a second

**Comment by pracucci** [comment]
Merging to continue work with #2281.
